scrapy/pipelines/images.py

```diff
@@ -127,15 +127,15 @@ def get_images(self, response, request, info):
             raise ImageException("Image too small (%dx%d < %dx%d)" %
                                  (width, height, self.min_width, self.min_height))
 
-        image, buf = self.convert_image(orig_image)
+        image, buf = self.convert_image(orig_image, BytesIO(response.body))
         yield path, image, buf
 
         for thumb_id, size in six.iteritems(self.thumbs):
             thumb_path = self.thumb_path(request, thumb_id, response=response, info=info)
-            thumb_image, thumb_buf = self.convert_image(image, size)
+            thumb_image, thumb_buf = self.convert_image(image, buf, size)
             yield thumb_path, thumb_image, thumb_buf
 
-    def convert_image(self, image, size=None):
+    def convert_image(self, image, response_body, size=None):
```

Comment:
Unfortunately this is backwards incompatible; there are image pipeline subclasses in a wild which override this method - they'll stop working because of signature change. Pipeline may also override a method which calls this method. Could you please make it backwards compatible? It seems something like this is needed:

* make response_body optional (None by default), make it a last argument
* return buf if response_body is not passed (i.e. disable the feature in this PR if a middleware uses deprecated signature).
* because user may override convert_image method, pipeline shouldn't always pass response_body - it should inspect convert_image method signature, and pass response_body only if this argument is present.
* issue a warning when response_body is not passed (ScrapyDeprecationWarning) - it seems two warnings are needed, one when convert_image is overridden in incompatible way, and another one is when convert_image is called from an overridden method in incompatible way.

Reply:
@kmike I have made changes, could you review this again?

Reply:
@anubhavp28 One thing you could do, since the goal is to support the previous API, is to, instead of modifying the tests to use the new API, keep the old tests and add new tests that test the new API.

You could also extend the old tests to ensure that a deprecation warning is used.

Reply:
@Gallaecio Not sure how to write tests to ensure that deprecation warnings are received. Currently, I am just counting the number of warnings. Would that be enough?
https://github.com/scrapy/scrapy/blob/881bade2c1b2d5842f7d0bdf2648455996f814b7/tests/test_pipeline_images.py#L111

Reply:
@anubhavp28 Please, check other usages of `catch_warnings` in the tests and you’ll see how to check the warning message as well. You could ensure that the warnings contain certain text to ensure that they are the expected warnings, and not some other warnings.